src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/step/rollover/AttemptRolloverStep.kt

```diff
@@ -99,18 +100,28 @@ class AttemptRolloverStep(
                 }
         ).toMap()
 
-        if (config.evaluateConditions(indexAgeTimeValue, numDocs, indexSize)) {
-            logger.info("$indexName rollover conditions evaluated to true [indexCreationDate=$indexCreationDate," +
-                    " numDocs=$numDocs, indexSize=${indexSize.bytes}]")
-            executeRollover(alias, conditions)
+        if (evaluateWriteIndex(alias)) {
+            if (config.evaluateConditions(indexAgeTimeValue, numDocs, indexSize)) {
+                logger.info("$indexName rollover conditions evaluated to true [indexCreationDate=$indexCreationDate," +
+                        " numDocs=$numDocs, indexSize=${indexSize.bytes}]")
+                executeRollover(alias, conditions)
+            } else {
+                stepStatus = StepStatus.CONDITION_NOT_MET
+                info = mapOf("message" to getAttemptingMessage(indexName), "conditions" to conditions)
+            }
         } else {
-            stepStatus = StepStatus.CONDITION_NOT_MET
-            info = mapOf("message" to getAttemptingMessage(indexName), "conditions" to conditions)
+            stepStatus = StepStatus.FAILED
+            info = mapOf("message" to getFailedNonWriteIndexMessage(indexName, alias))
         }
 
         return this
     }
 
+    private fun evaluateWriteIndex(alias: String): Boolean {
```

Comment:
The `aliases.get()` part here will block which we want to avoid. We can refactor this to be a suspend function and utilize the suspendUntil extension function.

This is also a network request which can throw an exception so we'll need to wrap it in a try/catch and handle the exception.

```kotlin
private suspend fun evaluateWriteIndex(alias: String): Boolean {
    try {
        val aliasesRequest = GetAliasesRequest().aliases(alias).indices(indexName)
        val aliasesResponse: GetAliasesResponse = client.admin().indices().suspendUntil { getAliases(aliasesRequest, it) 
        return aliasesResponse.aliases[indexName].firstOrNull()?.writeIndex() == true
    } catch (e: Exception) {
        handleException(e)
    }
}
```

This kind of also opens up the question, if the GetAliases request fails do we really want it to fail the whole Rollover action? Technically this will be called every time the managed index comes to life and checks whether it's time to rollover. So for some people that might rollover after a long time that means it's doing [time it takes to rollover] / [time in job_interval] requests that have the potential to fail for random network reasons. A user can of course set up some retry configuration on the rollover action to reduce permanent failures because of transient network failures, but perhaps having this be a CONDITION_NOT_MET that has a warning message instead? We don't have to tackle this in this PR, just something that perhaps we'll change in the future. For now we can have it treat it as a failure.

And lastly, there is a very slight gap in between the evaluateWriteIndex and rollover call that technically could allow your problem to still happen. Unlikely, but possible. Wondering if in future there is a better way to ensure it can't happen.

Reply:
https://www.elastic.co/guide/en/elasticsearch/reference/7.x/skipping-rollover.html

Would it be possible to implement the above way instead? In the case of OpenDistro ILM implementation, maybe set the "rolled_over" : true? This is the response that one gets from the Rollover API anyways.

Reply:
Hey @solomonsunder

To be clear you're saying a way for you to manually set rolled_over to be true?

I think in terms of automation there are two choices. If the above happens (write alias is pointing to a different index now) then we can either just accept this index was already rolled over (outside of ISM) and set rolled_over: true and just move on. Or because ISM definitely did not do it we can fail as it might mean there is some outside process doing something unexpectedly. I'm assuming we'll have a mix of wants from different users so perhaps the best choice is to just let the user choose by a rollover setting in the action.